---
title: "How to use UMAP transform on a single cell dataset (Seurat) using iterative Latent Semantic Indexing"
output: html_document
date: "2024-09-23"
always_allow_html: true
editor_options: 
  chunk_output_type: console
---

Note that this code was inspired by and adapted from: 

ArchR: An integrative and scalable software package for single-cell chromatin accessibility analysis
#' Jeffrey M. Granja, M. Ryan Corces, Sarah E. Pierce, S. Tansu Bagdatli, Hani Choudhry, Howard Y. Chang, William J. Greenleaf
#' doi: https://doi.org/10.1101/2020.04.28.066498



```{r, echo=FALSE, results='hide', message=FALSE, warning=FALSE}
rm(list=ls())
# Define dataset paths depending on your machine
if(grepl("^gizmo", Sys.info()["nodename"])){
  ROOT_DIR1<-"/fh/fast/furlan_s/experiments/MB_10X_5p/cds"
  ROOT_DIR2<-"/fh/fast/furlan_s/grp/data/ddata/BM_data"
  ROOT_DIR3<-"/fh/fast/furlan_s/experiments/LRPBMC/cds"
} else {
  ROOT_DIR1<-"/Users/sfurlan/Library/CloudStorage/OneDrive-SharedLibraries-FredHutchinsonCancerCenter/Furlan_Lab - General/experiments/MB_10X_5p/cds"
  ROOT_DIR2<-"/Users/sfurlan/Library/CloudStorage/OneDrive-SharedLibraries-FredHutchinsonCancerCenter/Furlan_Lab - General/datasets/Healthy_BM_greenleaf"
  ROOT_DIR3<-"/Users/sfurlan/Library/CloudStorage/OneDrive-SharedLibraries-FredHutchinsonCancerCenter/Furlan_Lab - General/experiments/LRPBMC/cds"
}
knitr::opts_chunk$set(message= FALSE, warning = FALSE)
```

## **Installing flscuts**

First, ensure you have the `devtools` R package installed, which allows you to install packages from GitHub. If `devtools` is installed, you can easily install using the following command:

```{r, eval=F}
devtools::install_github("furlan-lab/flscuts")
```

## **Loading data**

In this section, weâ€™ll load two Seurat objects, fix the celltypes so they harmonize, and create some colors.

```{r}
# Load required packages
suppressPackageStartupMessages({
  library(flscuts)
    library(Seurat)
  library(monocle3)
  library(viewmastR)
  library(ggplot2)
  library(scCustomize)
  library(tidyr)
})
```


#### Load reference dataset and make some colors
```{r}
seur <- readRDS(file.path(ROOT_DIR1, "240926_final_object.RDS"))
# Make some colors
cols1 <- as.character(pals::polychrome()[c(1,3:13)])
names(cols1) <- levels(factor(seur$celltype))
cols2 <- c("00_PROJECTOR"="grey90", cols1)

DimPlot_scCustom(seur, group.by = "celltype", colors_use = cols1)+theme_void()+ggtitle("Reference - aka Projector Dataset")
```

#### Load query dataset
```{r}
seu <-readRDS(file.path(ROOT_DIR3, "240926_5p_seu.RDS"))
DimPlot_scCustom(seu, group.by = "celltype", colors_use = cols1)+theme_void()+ggtitle("Query - aka Projectee Dataset")
```


## **Make sure labels are correct**

In this code, we show that the labels given to the reference and query cells are correct.

#### First the reference
```{r, fig.width = 12}
Idents(seur) <- seur$celltype
all_markers <- FindAllMarkers(object = seur) %>%
    Add_Pct_Diff() %>% dplyr::filter(pct_diff > 0.6)
    

top_markers <- Extract_Top_Markers(marker_dataframe = all_markers, num_genes = 5, named_vector = FALSE,
    make_unique = TRUE)

Clustered_DotPlot(seurat_object = seur, features = top_markers, colors_use_idents = cols1, k=11, plot_km_elbow = F)
```


#### And the query
```{r, fig.width = 12}
Idents(seu) <- seu$celltype
all_markers <- FindAllMarkers(object = seu) %>%
    Add_Pct_Diff() %>% dplyr::filter(pct_diff > 0.6)
    

top_markers <- Extract_Top_Markers(marker_dataframe = all_markers, num_genes = 5, named_vector = FALSE,
    make_unique = TRUE)

Clustered_DotPlot(seurat_object = seur, features = top_markers, colors_use_idents = cols1, k=11, plot_km_elbow = F)
```


## **Find shared features**

In this section we identify similarly variant genes that are common across the two datasets
```{r, fig.width = 6}
# Calculate and plot gene dispersion in query dataset
seu <- calculate_gene_dispersion(seu)
seu <- select_genes(seu, top_n = 10000, logmean_ul = -1, logmean_ll = -8)
plot_gene_dispersion(seu)
vgq <- get_selected_genes(seu)

# Repeat the process for the reference dataset
seur <- calculate_gene_dispersion(seur)
seur <- select_genes(seur, top_n = 10000, logmean_ul = -1, logmean_ll = -8)
plot_gene_dispersion(seur)
vgr <- get_selected_genes(seur)

# Find common genes
vg <- intersect(vgq, vgr)
```

## **Iterative Latent Sematic Indexing**

The overall goal is to use "iterative" (only 1 iteration) LSI with the same feature set on both datasets to create a reduction that can be used with umap transform

First, we run iterative LSI on the reference. We use an arbitrary number of LSI components and the common variant features found above.  Setting run_umap to true will enable this function to run, saving the UMAP model to the object.

```{r}
#iterative LSI
comps <- 25 #Number of PCs for clustering
seur<-iterative_LSI(seur, num_dim = comps, num_features = length(vg), resolution = c(1e-3), verbose = T, starting_features = vg, run_umap = T)

```


The UMAP performed on LSI looks like this.  Which is reasonable

```{r}
p1 <- DimPlot(seur, group.by = "celltype", cols = cols1)+theme_void()+ggtitle("Reference - aka Projector Dataset after LSI/UMAP")
p1
```


Next we take the "projectee", (aka the query data we want to project) and the "projector" aka reference and run the following function which creates an output containing the coordinates for how the cells from the projectee mapped using the same features+LSI/UMAP reduction procedure mapped the reference cells.  In this example, the cells we had previously seen 


```{r}

res <- project_data(projector = seur, projectee = seu, reduced_dim = "lsi", embedding = "umap")

```




```{r}
p2 <- plot_projection(res, seur, seu, projectee_col = "celltype")+scale_color_manual(values = cols2)+theme_void()+ guides(color = guide_legend(override.aes = list(size = 3)))
p2

```

Side by side this looks pretty good with the exception of Tregs, which get a bit lost.

```{r, fig.width=12}
cowplot::plot_grid(p1, p2)

```


## **Appendix**

```{r}
sessionInfo()
getwd()
```
