<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="flscuts">
<title>How to use UMAP transform on a single cell dataset (Seurat) using iterative Latent Semantic Indexing • flscuts</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/IBM_Plex_Sans-0.4.9/font.css" rel="stylesheet">
<link href="../deps/Fira_Mono-0.4.9/font.css" rel="stylesheet">
<link href="../deps/Alegreya_Sans_SC-0.4.9/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="How to use UMAP transform on a single cell dataset (Seurat) using iterative Latent Semantic Indexing">
<meta property="og:description" content="flscuts">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">flscuts</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Get Started</h6>
    <a class="dropdown-item" href="../articles/ProjectData.html">How to use UMAP transform on a single cell dataset (Seurat) using iterative Latent Semantic Indexing</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="http://twitter.com/scottfurlan" aria-label="Twitter">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/furlan-lab/viewmastR" aria-label="GitHub">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>How to use UMAP transform on a single cell dataset (Seurat) using iterative Latent Semantic Indexing</h1>
            
            <h4 data-toc-skip class="date">2024-09-23</h4>
      
      
      <div class="d-none name"><code>ProjectData.Rmd</code></div>
    </div>

    
    
<p>Note that this code was inspired by and adapted from:</p>
<p>ArchR: An integrative and scalable software package for single-cell
chromatin accessibility analysis #’ Jeffrey M. Granja, M. Ryan Corces,
Sarah E. Pierce, S. Tansu Bagdatli, Hani Choudhry, Howard Y. Chang,
William J. Greenleaf #’ doi: <a href="https://doi.org/10.1101/2020.04.28.066498" class="external-link uri">https://doi.org/10.1101/2020.04.28.066498</a></p>
<div class="section level2">
<h2 id="installing-flscuts">
<strong>Installing flscuts</strong><a class="anchor" aria-label="anchor" href="#installing-flscuts"></a>
</h2>
<p>First, ensure you have the <code>devtools</code> R package installed,
which allows you to install packages from GitHub. If
<code>devtools</code> is installed, you can easily install using the
following command:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"furlan-lab/flscuts"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="loading-data">
<strong>Loading data</strong><a class="anchor" aria-label="anchor" href="#loading-data"></a>
</h2>
<p>In this section, we’ll load two Seurat objects, fix the celltypes so
they harmonize, and create some colors.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load required packages</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="op">{</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://furlan-lab.github.io/flscuts/">flscuts</a></span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">monocle3</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://furlan-lab.github.io/viewmastR/" class="external-link">viewmastR</a></span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/samuel-marsh/scCustomize" class="external-link">scCustomize</a></span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org" class="external-link">tidyr</a></span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="load-reference-dataset-and-make-some-colors">Load reference dataset and make some colors<a class="anchor" aria-label="anchor" href="#load-reference-dataset-and-make-some-colors"></a>
</h4>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seur</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">ROOT_DIR1</span>, <span class="st">"240926_final_object.RDS"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># Make some colors</span></span>
<span><span class="va">cols1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fu">pals</span><span class="fu">::</span><span class="fu"><a href="https://kwstat.github.io/pals/reference/discrete.html" class="external-link">polychrome</a></span><span class="op">(</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">3</span><span class="op">:</span><span class="fl">13</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cols1</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">seur</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">cols2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"00_PROJECTOR"</span><span class="op">=</span><span class="st">"grey90"</span>, <span class="va">cols1</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://samuel-marsh.github.io/scCustomize/reference/DimPlot_scCustom.html" class="external-link">DimPlot_scCustom</a></span><span class="op">(</span><span class="va">seur</span>, group.by <span class="op">=</span> <span class="st">"celltype"</span>, colors_use <span class="op">=</span> <span class="va">cols1</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Reference - aka Projector Dataset"</span><span class="op">)</span></span></code></pre></div>
<p><img src="ProjectData_files/figure-html/unnamed-chunk-4-1.png" width="100%"></p>
</div>
<div class="section level4">
<h4 id="load-query-dataset">Load query dataset<a class="anchor" aria-label="anchor" href="#load-query-dataset"></a>
</h4>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu</span> <span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">ROOT_DIR3</span>, <span class="st">"240926_5p_seu.RDS"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://samuel-marsh.github.io/scCustomize/reference/DimPlot_scCustom.html" class="external-link">DimPlot_scCustom</a></span><span class="op">(</span><span class="va">seu</span>, group.by <span class="op">=</span> <span class="st">"celltype"</span>, colors_use <span class="op">=</span> <span class="va">cols1</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Query - aka Projectee Dataset"</span><span class="op">)</span></span></code></pre></div>
<p><img src="ProjectData_files/figure-html/unnamed-chunk-5-1.png" width="100%"></p>
</div>
</div>
<div class="section level2">
<h2 id="make-sure-labels-are-correct">
<strong>Make sure labels are correct</strong><a class="anchor" aria-label="anchor" href="#make-sure-labels-are-correct"></a>
</h2>
<p>In this code, we show that the labels given to the reference and
query cells are correct.</p>
<div class="section level4">
<h4 id="first-the-reference">First the reference<a class="anchor" aria-label="anchor" href="#first-the-reference"></a>
</h4>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">seur</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">seur</span><span class="op">$</span><span class="va">celltype</span></span>
<span><span class="va">all_markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html" class="external-link">FindAllMarkers</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">seur</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://samuel-marsh.github.io/scCustomize/reference/Add_Pct_Diff.html" class="external-link">Add_Pct_Diff</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">pct_diff</span> <span class="op">&gt;</span> <span class="fl">0.6</span><span class="op">)</span></span>
<span>    </span>
<span></span>
<span><span class="va">top_markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://samuel-marsh.github.io/scCustomize/reference/Extract_Top_Markers.html" class="external-link">Extract_Top_Markers</a></span><span class="op">(</span>marker_dataframe <span class="op">=</span> <span class="va">all_markers</span>, num_genes <span class="op">=</span> <span class="fl">5</span>, named_vector <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    make_unique <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://samuel-marsh.github.io/scCustomize/reference/Clustered_DotPlot.html" class="external-link">Clustered_DotPlot</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">seur</span>, features <span class="op">=</span> <span class="va">top_markers</span>, colors_use_idents <span class="op">=</span> <span class="va">cols1</span>, k<span class="op">=</span><span class="fl">11</span>, plot_km_elbow <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre></div>
<p><img src="ProjectData_files/figure-html/unnamed-chunk-6-1.png" width="100%"></p>
</div>
<div class="section level4">
<h4 id="and-the-query">And the query<a class="anchor" aria-label="anchor" href="#and-the-query"></a>
</h4>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">seu</span><span class="op">$</span><span class="va">celltype</span></span>
<span><span class="va">all_markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html" class="external-link">FindAllMarkers</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">seu</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://samuel-marsh.github.io/scCustomize/reference/Add_Pct_Diff.html" class="external-link">Add_Pct_Diff</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">pct_diff</span> <span class="op">&gt;</span> <span class="fl">0.6</span><span class="op">)</span></span>
<span>    </span>
<span></span>
<span><span class="va">top_markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://samuel-marsh.github.io/scCustomize/reference/Extract_Top_Markers.html" class="external-link">Extract_Top_Markers</a></span><span class="op">(</span>marker_dataframe <span class="op">=</span> <span class="va">all_markers</span>, num_genes <span class="op">=</span> <span class="fl">5</span>, named_vector <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    make_unique <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://samuel-marsh.github.io/scCustomize/reference/Clustered_DotPlot.html" class="external-link">Clustered_DotPlot</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">seur</span>, features <span class="op">=</span> <span class="va">top_markers</span>, colors_use_idents <span class="op">=</span> <span class="va">cols1</span>, k<span class="op">=</span><span class="fl">11</span>, plot_km_elbow <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre></div>
<p><img src="ProjectData_files/figure-html/unnamed-chunk-7-1.png" width="100%"></p>
</div>
</div>
<div class="section level2">
<h2 id="find-shared-features">
<strong>Find shared features</strong><a class="anchor" aria-label="anchor" href="#find-shared-features"></a>
</h2>
<p>In this section we identify similarly variant genes that are common
across the two datasets</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Calculate and plot gene dispersion in query dataset</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://furlan-lab.github.io/viewmastR/reference/calculate_gene_dispersion.html" class="external-link">calculate_gene_dispersion</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://furlan-lab.github.io/viewmastR/reference/select_genes.html" class="external-link">select_genes</a></span><span class="op">(</span><span class="va">seu</span>, top_n <span class="op">=</span> <span class="fl">10000</span>, logmean_ul <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>, logmean_ll <span class="op">=</span> <span class="op">-</span><span class="fl">8</span><span class="op">)</span></span>
<span><span class="fu">plot_gene_dispersion</span><span class="op">(</span><span class="va">seu</span><span class="op">)</span></span></code></pre></div>
<p><img src="ProjectData_files/figure-html/unnamed-chunk-8-1.png" width="100%"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vgq</span> <span class="op">&lt;-</span> <span class="fu">get_selected_genes</span><span class="op">(</span><span class="va">seu</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Repeat the process for the reference dataset</span></span>
<span><span class="va">seur</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://furlan-lab.github.io/viewmastR/reference/calculate_gene_dispersion.html" class="external-link">calculate_gene_dispersion</a></span><span class="op">(</span><span class="va">seur</span><span class="op">)</span></span>
<span><span class="va">seur</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://furlan-lab.github.io/viewmastR/reference/select_genes.html" class="external-link">select_genes</a></span><span class="op">(</span><span class="va">seur</span>, top_n <span class="op">=</span> <span class="fl">10000</span>, logmean_ul <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>, logmean_ll <span class="op">=</span> <span class="op">-</span><span class="fl">8</span><span class="op">)</span></span>
<span><span class="fu">plot_gene_dispersion</span><span class="op">(</span><span class="va">seur</span><span class="op">)</span></span></code></pre></div>
<p><img src="ProjectData_files/figure-html/unnamed-chunk-8-2.png" width="100%"></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vgr</span> <span class="op">&lt;-</span> <span class="fu">get_selected_genes</span><span class="op">(</span><span class="va">seur</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Find common genes</span></span>
<span><span class="va">vg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">vgq</span>, <span class="va">vgr</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="iterative-latent-sematic-indexing">
<strong>Iterative Latent Sematic Indexing</strong><a class="anchor" aria-label="anchor" href="#iterative-latent-sematic-indexing"></a>
</h2>
<p>The overall goal is to use “iterative” (only 1 iteration) LSI with
the same feature set on both datasets to create a reduction that can be
used with umap transform</p>
<p>First, we run iterative LSI on the reference. We use an arbitrary
number of LSI components and the common variant features found above.
Setting run_umap to true will enable this function to run, saving the
UMAP model to the object.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#iterative LSI</span></span>
<span><span class="va">comps</span> <span class="op">&lt;-</span> <span class="fl">25</span> <span class="co">#Number of PCs for clustering</span></span>
<span><span class="va">seur</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/iterative_LSI.html">iterative_LSI</a></span><span class="op">(</span><span class="va">seur</span>, num_dim <span class="op">=</span> <span class="va">comps</span>, num_features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">vg</span><span class="op">)</span>, resolution <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1e-3</span><span class="op">)</span>, verbose <span class="op">=</span> <span class="cn">T</span>, starting_features <span class="op">=</span> <span class="va">vg</span>, run_umap <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of nodes: 11290</span></span>
<span><span class="co">## Number of edges: 448690</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Running Louvain algorithm...</span></span>
<span><span class="co">## Maximum modularity in 10 random starts: 0.9996</span></span>
<span><span class="co">## Number of communities: 3</span></span>
<span><span class="co">## Elapsed time: 0 seconds</span></span></code></pre>
<p>The UMAP performed on LSI looks like this. Which is reasonable</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seur</span>, group.by <span class="op">=</span> <span class="st">"celltype"</span>, cols <span class="op">=</span> <span class="va">cols1</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Reference - aka Projector Dataset after LSI/UMAP"</span><span class="op">)</span></span>
<span><span class="va">p1</span></span></code></pre></div>
<p><img src="ProjectData_files/figure-html/unnamed-chunk-10-1.png" width="100%"></p>
<p>Next we take the “projectee”, (aka the query data we want to project)
and the “projector” aka reference and run the following function which
creates an output containing the coordinates for how the cells from the
projectee mapped using the same features+LSI/UMAP reduction procedure
mapped the reference cells. In this example, the cells we had previously
seen</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/project_data.html">project_data</a></span><span class="op">(</span>projector <span class="op">=</span> <span class="va">seur</span>, projectee <span class="op">=</span> <span class="va">seu</span>, reduced_dim <span class="op">=</span> <span class="st">"lsi"</span>, embedding <span class="op">=</span> <span class="st">"umap"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_projection.html">plot_projection</a></span><span class="op">(</span><span class="va">res</span>, <span class="va">seur</span>, <span class="va">seu</span>, projectee_col <span class="op">=</span> <span class="st">"celltype"</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">cols2</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>color <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p2</span></span></code></pre></div>
<p><img src="ProjectData_files/figure-html/unnamed-chunk-12-1.png" width="100%"></p>
<p>Side by side this looks pretty good with the exception of Tregs,
which get a bit lost.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span><span class="op">)</span></span></code></pre></div>
<p><img src="ProjectData_files/figure-html/unnamed-chunk-13-1.png" width="100%"></p>
</div>
<div class="section level2">
<h2 id="appendix">
<strong>Appendix</strong><a class="anchor" aria-label="anchor" href="#appendix"></a>
</h2>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">## Platform: x86_64-apple-darwin20 (64-bit)</span></span>
<span><span class="co">## Running under: macOS Sonoma 14.6.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: America/Los_Angeles</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] Matrix_1.6-5                tidyr_1.3.1                </span></span>
<span><span class="co">##  [3] scCustomize_2.1.2           ggplot2_3.5.0              </span></span>
<span><span class="co">##  [5] viewmastR_0.2.3             monocle3_1.4.3             </span></span>
<span><span class="co">##  [7] SingleCellExperiment_1.22.0 SummarizedExperiment_1.30.2</span></span>
<span><span class="co">##  [9] GenomicRanges_1.52.1        GenomeInfoDb_1.41.1        </span></span>
<span><span class="co">## [11] IRanges_2.34.1              S4Vectors_0.38.2           </span></span>
<span><span class="co">## [13] MatrixGenerics_1.12.3       matrixStats_1.2.0          </span></span>
<span><span class="co">## [15] Biobase_2.60.0              BiocGenerics_0.46.0        </span></span>
<span><span class="co">## [17] Seurat_5.0.3                SeuratObject_5.0.1         </span></span>
<span><span class="co">## [19] sp_2.1-3                    flscuts_0.1.0              </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] fs_1.6.3                  spatstat.sparse_3.0-3    </span></span>
<span><span class="co">##   [3] bitops_1.0-7              lubridate_1.9.3          </span></span>
<span><span class="co">##   [5] RcppMsgPack_0.2.3         httr_1.4.7               </span></span>
<span><span class="co">##   [7] RColorBrewer_1.1-3        doParallel_1.0.17        </span></span>
<span><span class="co">##   [9] backports_1.4.1           tools_4.3.1              </span></span>
<span><span class="co">##  [11] sctransform_0.4.1         utf8_1.2.4               </span></span>
<span><span class="co">##  [13] R6_2.5.1                  lazyeval_0.2.2           </span></span>
<span><span class="co">##  [15] uwot_0.1.16               GetoptLong_1.0.5         </span></span>
<span><span class="co">##  [17] withr_3.0.0               gridExtra_2.3            </span></span>
<span><span class="co">##  [19] progressr_0.14.0          cli_3.6.2                </span></span>
<span><span class="co">##  [21] textshaping_0.3.7         Cairo_1.6-2              </span></span>
<span><span class="co">##  [23] spatstat.explore_3.2-6    fastDummies_1.7.3        </span></span>
<span><span class="co">##  [25] labeling_0.4.3            sass_0.4.9               </span></span>
<span><span class="co">##  [27] spatstat.data_3.0-4       ggridges_0.5.6           </span></span>
<span><span class="co">##  [29] pbapply_1.7-2             pkgdown_2.0.7            </span></span>
<span><span class="co">##  [31] systemfonts_1.0.6         foreign_0.8-86           </span></span>
<span><span class="co">##  [33] dichromat_2.0-0.1         parallelly_1.37.1        </span></span>
<span><span class="co">##  [35] maps_3.4.2                limma_3.56.2             </span></span>
<span><span class="co">##  [37] pals_1.8                  rstudioapi_0.15.0        </span></span>
<span><span class="co">##  [39] generics_0.1.3            shape_1.4.6.1            </span></span>
<span><span class="co">##  [41] gtools_3.9.5              ica_1.0-3                </span></span>
<span><span class="co">##  [43] spatstat.random_3.2-3     dplyr_1.1.4              </span></span>
<span><span class="co">##  [45] ggbeeswarm_0.7.2          fansi_1.0.6              </span></span>
<span><span class="co">##  [47] abind_1.4-5               lifecycle_1.0.4          </span></span>
<span><span class="co">##  [49] yaml_2.3.8                edgeR_3.42.4             </span></span>
<span><span class="co">##  [51] snakecase_0.11.1          recipes_1.0.10           </span></span>
<span><span class="co">##  [53] Rtsne_0.17                paletteer_1.6.0          </span></span>
<span><span class="co">##  [55] grid_4.3.1                promises_1.2.1           </span></span>
<span><span class="co">##  [57] crayon_1.5.2              miniUI_0.1.1.1           </span></span>
<span><span class="co">##  [59] lattice_0.22-5            beachmat_2.16.0          </span></span>
<span><span class="co">##  [61] cowplot_1.1.3             mapproj_1.2.11           </span></span>
<span><span class="co">##  [63] pillar_1.9.0              knitr_1.45               </span></span>
<span><span class="co">##  [65] ComplexHeatmap_2.16.0     rjson_0.2.21             </span></span>
<span><span class="co">##  [67] boot_1.3-28.1             future.apply_1.11.1      </span></span>
<span><span class="co">##  [69] codetools_0.2-19          leiden_0.4.3.1           </span></span>
<span><span class="co">##  [71] glue_1.7.0                data.table_1.15.2        </span></span>
<span><span class="co">##  [73] vctrs_0.6.5               png_0.1-8                </span></span>
<span><span class="co">##  [75] spam_2.10-0               gtable_0.3.4             </span></span>
<span><span class="co">##  [77] rematch2_2.1.2            assertthat_0.2.1         </span></span>
<span><span class="co">##  [79] cachem_1.0.8              gower_1.0.1              </span></span>
<span><span class="co">##  [81] xfun_0.42                 prodlim_2023.08.28       </span></span>
<span><span class="co">##  [83] S4Arrays_1.2.0            mime_0.12                </span></span>
<span><span class="co">##  [85] tidygraph_1.3.1           survival_3.5-7           </span></span>
<span><span class="co">##  [87] timeDate_4032.109         pbmcapply_1.5.1          </span></span>
<span><span class="co">##  [89] iterators_1.0.14          hardhat_1.3.1            </span></span>
<span><span class="co">##  [91] lava_1.8.0                ellipsis_0.3.2           </span></span>
<span><span class="co">##  [93] fitdistrplus_1.1-11       ipred_0.9-14             </span></span>
<span><span class="co">##  [95] ROCR_1.0-11               nlme_3.1-164             </span></span>
<span><span class="co">##  [97] bit64_4.0.5               RcppAnnoy_0.0.22         </span></span>
<span><span class="co">##  [99] bslib_0.6.1               irlba_2.3.5.1            </span></span>
<span><span class="co">## [101] rpart_4.1.23              vipor_0.4.7              </span></span>
<span><span class="co">## [103] KernSmooth_2.23-22        colorspace_2.1-0         </span></span>
<span><span class="co">## [105] Hmisc_5.1-2               nnet_7.3-19              </span></span>
<span><span class="co">## [107] ggrastr_1.0.2             tidyselect_1.2.1         </span></span>
<span><span class="co">## [109] bit_4.0.5                 compiler_4.3.1           </span></span>
<span><span class="co">## [111] htmlTable_2.4.2           BiocNeighbors_1.18.0     </span></span>
<span><span class="co">## [113] hdf5r_1.3.11              desc_1.4.3               </span></span>
<span><span class="co">## [115] DelayedArray_0.26.7       plotly_4.10.4            </span></span>
<span><span class="co">## [117] checkmate_2.3.1           scales_1.3.0             </span></span>
<span><span class="co">## [119] lmtest_0.9-40             stringr_1.5.1            </span></span>
<span><span class="co">## [121] digest_0.6.35             goftest_1.2-3            </span></span>
<span><span class="co">## [123] presto_1.0.0              spatstat.utils_3.1-0     </span></span>
<span><span class="co">## [125] minqa_1.2.6               rmarkdown_2.26           </span></span>
<span><span class="co">## [127] XVector_0.40.0            htmltools_0.5.7          </span></span>
<span><span class="co">## [129] pkgconfig_2.0.3           base64enc_0.1-3          </span></span>
<span><span class="co">## [131] lme4_1.1-35.1             sparseMatrixStats_1.12.2 </span></span>
<span><span class="co">## [133] highr_0.10                fastmap_1.1.1            </span></span>
<span><span class="co">## [135] rlang_1.1.3               GlobalOptions_0.1.2      </span></span>
<span><span class="co">## [137] htmlwidgets_1.6.4         UCSC.utils_1.1.0         </span></span>
<span><span class="co">## [139] shiny_1.8.0               DelayedMatrixStats_1.22.6</span></span>
<span><span class="co">## [141] farver_2.1.1              jquerylib_0.1.4          </span></span>
<span><span class="co">## [143] zoo_1.8-12                jsonlite_1.8.8           </span></span>
<span><span class="co">## [145] BiocParallel_1.34.2       ModelMetrics_1.2.2.2     </span></span>
<span><span class="co">## [147] BiocSingular_1.16.0       RCurl_1.98-1.14          </span></span>
<span><span class="co">## [149] magrittr_2.0.3            Formula_1.2-5            </span></span>
<span><span class="co">## [151] GenomeInfoDbData_1.2.10   dotCall64_1.1-1          </span></span>
<span><span class="co">## [153] patchwork_1.2.0           munsell_0.5.0            </span></span>
<span><span class="co">## [155] Rcpp_1.0.12               viridis_0.6.5            </span></span>
<span><span class="co">## [157] reticulate_1.35.0         pROC_1.18.5              </span></span>
<span><span class="co">## [159] stringi_1.8.3             miloR_1.8.1              </span></span>
<span><span class="co">## [161] ggraph_2.2.1              zlibbioc_1.46.0          </span></span>
<span><span class="co">## [163] MASS_7.3-60.0.1           plyr_1.8.9               </span></span>
<span><span class="co">## [165] parallel_4.3.1            listenv_0.9.1            </span></span>
<span><span class="co">## [167] ggrepel_0.9.5             forcats_1.0.0            </span></span>
<span><span class="co">## [169] deldir_2.0-4              graphlayouts_1.1.0       </span></span>
<span><span class="co">## [171] splines_4.3.1             tensor_1.5               </span></span>
<span><span class="co">## [173] circlize_0.4.16           locfit_1.5-9.8           </span></span>
<span><span class="co">## [175] igraph_2.0.3              spatstat.geom_3.2-9      </span></span>
<span><span class="co">## [177] RcppHNSW_0.6.0            reshape2_1.4.4           </span></span>
<span><span class="co">## [179] ScaledMatrix_1.8.1        evaluate_0.23            </span></span>
<span><span class="co">## [181] ggprism_1.0.4             nloptr_2.0.3             </span></span>
<span><span class="co">## [183] foreach_1.5.2             tweenr_2.0.3             </span></span>
<span><span class="co">## [185] httpuv_1.6.14             RANN_2.6.1               </span></span>
<span><span class="co">## [187] purrr_1.0.2               polyclip_1.10-6          </span></span>
<span><span class="co">## [189] future_1.33.1             clue_0.3-65              </span></span>
<span><span class="co">## [191] scattermore_1.2           ggforce_0.4.2            </span></span>
<span><span class="co">## [193] janitor_2.2.0             rsvd_1.0.5               </span></span>
<span><span class="co">## [195] xtable_1.8-4              RSpectra_0.16-1          </span></span>
<span><span class="co">## [197] later_1.3.2               class_7.3-22             </span></span>
<span><span class="co">## [199] viridisLite_0.4.2         ragg_1.3.0               </span></span>
<span><span class="co">## [201] tibble_3.2.1              memoise_2.0.1            </span></span>
<span><span class="co">## [203] beeswarm_0.4.0            cluster_2.1.6            </span></span>
<span><span class="co">## [205] timechange_0.3.0          globals_0.16.3           </span></span>
<span><span class="co">## [207] caret_6.0-94</span></span></code></pre>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "/Users/sfurlan/develop/flscuts/vignettes"</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by <a href="https://furlan-lab.github.io" class="external-link">Scott Furlan</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
